### PR TITLE
ista pgd

### PR DESCRIPTION
**Pull request by EnLAI111**
I have added a random diagonal matrix A in the generator of dataset, ```dataset/simulated.py```. 
When A is full of random elements, the speed of synthesic solver ```ISTA``` will become far more slow than that of analytic solver ```PGD```, which is raisonable.

### CODE REVIEW

File: objective.py [hunk 1]
```diff
@@ -8,14 +8,17 @@ class Objective(BaseObjective):
...
+        R = self.y - self.A @ u
+        return .5 * R @ R + self.reg*self.reg_max*(abs(np.diff(u)).sum())
 
     def to_dict(self):
-        return dict(reg=self.reg, reg_max=self.reg_max, y=self.y)
+        S = np.sum(self.A, axis=1)
+        c = S.dot(self.y)/(S.dot(S))*np.eye(len(self.y), dtype=int)
+        reg_max = np.max(abs((self.A).T.dot(self.A.dot(c) - self.y)))
+        self.reg_max = reg_max
+        return dict(A=self.A, reg=self.reg, reg_max=self.reg_max, y=self.y)
```
Comment (tomMoral) [inline objective.py:24]: Just pass `reg` here as `self.reg * reg_max`.

**Review by tomMoral — COMMENTED** [review]
A few comments but once you are done, feel free to merge this PR and move on to the next solver.

File: datasets/simulated.py [hunk 2]
```diff
@@ -10,27 +10,35 @@ class Dataset(BaseDataset):
...
+    def set_A(self, rng):
+        if(self.type_A == 'diagonal'):
+            A = np.diag(rng.random(self.K))
+        elif(self.type_A == 'triangulaire'):
+            A = np.triu(rng.rand(self.K, self.K))
+        elif(self.type_A == 'random'):
+            A = rng.rand(self.K, self.K)
+        else:
+            A = np.eye(self.K, dtype=int)
+        return A
```
Comment (agramfort) [inline datasets/simulated.py:36]:
```suggestion
        if self.type_A == 'diagonal':
            A = np.diag(rng.random(self.K))
        elif self.type_A == 'triangular':
            A = np.triu(rng.rand(self.K, self.K))
        elif self.type_A == 'random':
            A = rng.randn(self.K, self.K)
        else:
            A = np.eye(self.K, dtype=float)
        return A
```

File: datasets/simulated.py [hunk 3]
```diff
@@ -10,27 +10,35 @@ class Dataset(BaseDataset):
     # the cross product for each key in the dictionary.
     # cos + bruit ~ N(mu, sigma)
     parameters = {
-        'sigma': np.linspace(0.1, 0.5, 3).round(1),
+        'sigma': [0.1],
         'mu': [0],
-        'K': [20]}
+        'K': [20],
+        'type_A': ['identity', 'diagonal', 'triangulaire', 'random']}
```
Comment (agramfort) [inline datasets/simulated.py:16]:
```suggestion
        'type_A': ['identity', 'diagonal', 'triangular', 'random']}
```

**Comment by agramfort** [comment]
@EnLAI111 you ignored my comments above apparently

**Pull request merged**